libnmstate/nmstate.py

```diff
@@ -0,0 +1,51 @@
+#
+# Copyright (c) 2018-2020 Red Hat, Inc.
+#
+# This file is part of nmstate
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 2.1 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+#
+
+from contextlib import contextmanager
+
+from libnmstate import validator
+from libnmstate.nm import NetworkManagerPlugin
+from libnmstate.schema import DNS
+from libnmstate.schema import Interface
+from libnmstate.schema import Route
+from libnmstate.schema import RouteRule
+
+
+@contextmanager
+def plugin_context():
+    nm_plugin = NetworkManagerPlugin()
+    try:
+        yield nm_plugin
+    finally:
+        nm_plugin.unload()
+
+
+def show_with_plugin(plugin, include_status_data=None):
```

Comment:
This break the API.

Currently the `show()` function does not require user to provide plugin.

The file `nmstate.py` is preparation of plugin design:

```python
nmstate = NmState() # Load the buildin plugins.
nmstate.load_python_plugin(OvsNmstatePlugin())
nmstate.load_varlink_plugin("unix:/tmp/info.grisge.nispor.varlink")
state = nmstate.show()
nmstate.apply(state)
nmstate.unload_all()
```

As the subject of PR indicate, this PR only remove the global `NM.Client` used by top level
modules.

Reply:
Thanks for sharing how you see this evolving, I think it is a bit clearer now.

My concern here is that things get centralized. It is fine to centralize interfaces but not implementation.
I was not intending to suggest exposing the `show(plugin)` as an external api. I was suggesting to keep the filling of the report in a common generic place, where it takes the data from the individual plugin (which implements the plugin interface) and fills the report.

The current name of this module (`nmstate`) is odd, but maybe I should let it roll out and see how it plays out such that it will make more sense later.

Reply:
The `nmstate.py` will entry point of this whole module once `netinfo.py` and `netapplier.py` been cleaned up from plugin dependent.